### PR TITLE
ENH: Add broadcast support to Generator.multinomial

### PR DESCRIPTION
**Pull request by bashtage**
Add broadcasting support to multinomial

xref #15201

### CODE REVIEW

**Comment by yoavram** [comment]
I’d be happy to be a test user when this is ready

**Comment by mattip** [comment]
LGTM, but there is a merge conflict.

**Comment by bashtage** [comment]
@mattip Is this fail due to setuptools/distutils expected?

**Comment by charris** [comment]
> Is this fail due to setuptools/distutils expected?

Should be fixed in master, we don't allow the troublesome version of setuptools. I'll try a close/reopen, if that doesn't work you will probably need to rebase.

**Comment by bashtage** [comment]
Funny thing is that this was just rebased on master when the fails started.

**Comment by mattip** [comment]
It seems [setuptools 49.2.1 is out](https://dev.azure.com/numpy/numpy/_build/results?buildId=11801&view=logs&j=f49a0588-9972-5862-0c63-6d346787e7be&t=48451c78-bd0a-52bc-6e33-224066a9f4b3&l=17) but issue pypa/setuptools#2259 is not fixed. We will have to pin `setuptools<49.2.0`.

**Comment by mattip** [comment]
xref pr gh-16993

**Review by WarrenWeckesser — COMMENTED** [review]
I made a few suggestions, but I haven't carefully reviewed the implementation of the broadcasting.  I did run into problems with a couple edge cases: `rng.multinomial([1, 2], [])` raises `ZeroDivisionError: division by zero`, which is a strange error to get for that input, and more seriously, `rng.multinomial(5, [])` crashes Python with `Floating point exception: 8` on OSX and `Floating point exception (core dumped)` on Linux.

**Comment by bashtage** [comment]
I think I have addressed all of the issues you identified @WarrenWeckesser . I also added checks for edge cases with empty arrays.

File: numpy/random/_generator.pyx [hunk 1]
```diff
@@ -3567,24 +3567,34 @@ cdef class Generator:
...
         Returns
         -------
         out : ndarray
-            The drawn samples, of shape *size*, if that was provided.  If not,
-            the shape is ``(N,)``.
+            The drawn samples, of shape size, if provided. When size is
+            provided, the output shape is size + (p,)  If not specified,
+            the shape is determined by the broadcast shape of ``n`` and
+            ``pvals``, ``(b0, b1, ..., bq)`` augmented with the dimension of
+            the multiconial, ``p``, so that that output shape is
```
[the remark this answers is not in the document]
Comment (bashtage) [reply numpy/random/_generator.pyx]: Fixed. Thanks.

File: numpy/random/_generator.pyx [hunk 2]
```diff
@@ -3567,24 +3567,34 @@ cdef class Generator:
...
+            ``(k0, k1, ..., kn, p)``. Each element ``pvals[i,j,...,:]`` must
+            sum to 1 (however, the last element is always assumed to account
+            for the remaining probability, as long as
+            ``sum(pvals[i,j,...,:-1]) <= 1)``.
         size : int or tuple of ints, optional
             Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
-            ``m * n * k`` samples are drawn.  Default is None, in which case a
-            single value is returned.
+            ``m * n * k`` samples are drawn.  Default is None where the output
+            size is determined by the broadcast shape of ``n`` and ``pvals``,
```
[the remark this answers is not in the document]
Comment (bashtage) [reply numpy/random/_generator.pyx]: I have tried to reexplain the broadcast semantics; might still be a bit dense.

File: numpy/random/tests/test_generator_mt19937.py [hunk 3]
```diff
@@ -2349,6 +2343,61 @@ def test_multinomial(self):
...
+    @pytest.mark.parametrize("n", [10,
+                                   np.array([10, 10]),
+                                   np.array([[[10]], [[10]]])
+                                   ]
+                             )
+    def test_multinomial_pval_broadcast(self, n):
+        random = Generator(MT19937(self.seed))
+        pvals = np.array([1 / 4] * 4)
+        actual = random.multinomial(n, pvals)
+        assert actual.shape == np.broadcast(n, 1).shape + (4,)
```
Comment (WarrenWeckesser) [inline numpy/random/tests/test_generator_mt19937.py:2369]: The significance of that `1` might not be apparent; here and below, I think we can make this check clearer by using the actual shape of `pvals`, by using `broadcast_shapes`:
```suggestion
        assert actual.shape == (np.broadcast_shapes(n.shape, pvals.shape[:-1])
                                + (4,))
```
(The weird line break is for PEP 8.)

File: numpy/random/tests/test_generator_mt19937.py [hunk 4]
```diff
@@ -2349,6 +2343,61 @@ def test_multinomial(self):
...
+        pvals = np.vstack([pvals, pvals])
+        actual = random.multinomial(n, pvals)
+        assert actual.shape == np.broadcast(n, np.ones(2)).shape + (4,)
```
Comment (WarrenWeckesser) [inline numpy/random/tests/test_generator_mt19937.py:2372]:
```suggestion
        assert actual.shape == (np.broadcast_shapes(n.shape, pvals.shape[:-1])
                                + (4,))
```

File: numpy/random/tests/test_generator_mt19937.py [hunk 5]
```diff
@@ -2349,6 +2343,61 @@ def test_multinomial(self):
...
+
+        pvals = np.vstack([[pvals], [pvals]])
+        actual = random.multinomial(n, pvals)
+        expected_shape = np.broadcast(n, np.ones((2, 2))).shape
```
Comment (WarrenWeckesser) [inline numpy/random/tests/test_generator_mt19937.py:2376]:
```suggestion
        expected_shape = np.broadcast_shapes(n.shape, pvals.shape[:-1])
```

**Review by WarrenWeckesser — CHANGES_REQUESTED** [review]
@bashtage, sorry this "dropped off my radar" (as they euphemistically say) for *so* long.
I made a few small suggestions, otherwise looks good.

File: numpy/random/tests/test_generator_mt19937.py (hunk 4 above)
Comment (bashtage) [reply numpy/random/tests/test_generator_mt19937.py:2372]: Done.

File: numpy/random/tests/test_generator_mt19937.py (hunk 3 above)
Comment (bashtage) [reply numpy/random/tests/test_generator_mt19937.py:2369]: Done.

File: numpy/random/_generator.pyx [hunk 6]
```diff
@@ -3675,24 +3675,35 @@ cdef class Generator:
...
+            ``sum(pvals[..., :-1], axis=-1) <= 1.0``. Must have at least 1
+            dimension where pvals.shape[-1] > 0.
         size : int or tuple of ints, optional
             Output shape.  If the given shape is, e.g., ``(m, n, k)``, then
-            ``m * n * k`` samples are drawn.  Default is None, in which case a
-            single value is returned.
+            ``m * n * k`` samples are drawn each with ``p`` elements. Default
+            is None where the output size is determined by the broadcast shape
+            of ``n`` and all by the final dimension of ``pvals``, which is
+            denoted as ``b=(b0, b1, ..., bq)`` be this size. If size is not None,
```
[the remark this answers is not in the document]
Comment (bashtage) [reply numpy/random/_generator.pyx:3690]: Done.

File: numpy/random/_generator.pyx [hunk 7]
```diff
@@ -3744,47 +3787,79 @@ cdef class Generator:
...
+                                             np.NPY_ARRAY_ALIGNED |
+                                             np.NPY_ARRAY_C_CONTIGUOUS)
+        parr = <np.ndarray>np.PyArray_FROM_OTF(pvals,
+                                               np.NPY_DOUBLE,
+                                               np.NPY_ARRAY_ALIGNED |
+                                               np.NPY_ARRAY_C_CONTIGUOUS)
+        ndim = parr.ndim
+        d = parr.shape[ndim - 1] if ndim >= 1 else 0
+        if d == 0:
+            raise ValueError("pvals must have at least 1 dimension with shape[-1] > 0.")
```
[the remark this answers is not in the document]
Comment (bashtage) [reply numpy/random/_generator.pyx:3810]: Changed to reflect comment.

**Review by WarrenWeckesser — APPROVED** [review]
Thanks @bashtage!

**Comment by bashtage** [comment]
Thanks @WarrenWeckesser

**Pull request merged**